Caching, Survey Details & Accept/Decline functionality

sophiabersani opened the pull request:
Implements many related tasks that allow the user to have control over the displayed surveys.

Major Changes:
- Exposes Realm.io to the rest of the app to allow for local saving.
- Adds a SurveyDetails view using cached data.
- Adds methods and styled buttons to accept/decline surveys, saved to Realm.
- Changes SurveyList to feed off stored Realm data instead of Parse to allow for dynamic list items.

Improvements:
- Performance: FormPage no longer needs to load Parse objects every time it's mounted, it uses cached data.
- Performance: API now caches forms and questions ahead of time for the surveys that are fetched.
- Changed SurveyListItem to allow for more than 2 states: (pending, accepted, declined)

On Hold/Blocked:
- SurveyDetailsPage is missing Organization names. That data is not being created by the admin end.
- Still only have support for a single form on click, this behavior should be looked into after we are able to generate background notifications to guide users to these forms.

Commit:
Adds Survey Detail view

Commit:
Adds styled accept/decline buttons

Commit:
Adds styled icons to survey detail buttons

Commit:
Allows user to view forms via the Survey Details page

Commit:
Exposes Realm.io to the rest of the app

Realm was being initialized and limited to FormPage component only.

Commit:
Fixing styling issue with Survey Details buttons

Commit:
Caches Surveys, Forms, and Questions to be used in detail views.

Commit:
Adds support for device-cached data in SurveyDetails view

Commit:
Adds support for Realm cache in SurveyList page

Commit:
Adds Realm cache support to Survey Items and Forms

Commit:
Merge branch 'master' into survey-details

# Conflicts:
#	js/router/SharedNavigator.js
#	js/views/FormPage.js

Commit:
Cleans some old code form merger.

Commit:
Adds error message associated with missing questions

dan-nyanko commented:
@gbersani 

I haven't looked into the code, but upon a fresh install of the app with this branch I consistently get multiple errors.

![image](https://cloud.githubusercontent.com/assets/12954045/15430134/fceb991a-1e71-11e6-8ac1-ff692bc4f45e.png)

dan-nyanko commented:
The survey list only has one item and its missing a title.

![image](https://cloud.githubusercontent.com/assets/12954045/15430249/7acf5aa6-1e72-11e6-9e8e-9d161ed014bc.png)

Upon selecting the survey, the details page appears but is missing data.

![image](https://cloud.githubusercontent.com/assets/12954045/15430310/c43ee42c-1e72-11e6-994e-f7091e22248f.png)

Pressing accept give the following error.

![image](https://cloud.githubusercontent.com/assets/12954045/15430328/da88bc08-1e72-11e6-80f3-918f206c81a4.png)

dan-nyanko commented on js/api/Forms.js:
```diff
@@ -1,18 +1,38 @@
   if (surveyFormRelations) {
     surveyFormRelations.query().find({
       success: function(results) {
         storeForms(results)
```
Why are we placing data in realm and within the data/Store.js arrays?

dan-nyanko commented on js/api/Questions.js:
```diff
@@ -1,13 +1,36 @@
   const formQuestionRelations = form.get('questions')
   formQuestionRelations.query().find({
     success: function(results) {
       storeQuestions(results)
```
Same question as above

sophiabersani commented:
There's no need to keep them in the Store anymore. I should clean that out.
As for the error: That is very odd, I will look into what might be causing this, but I am having a hard time reproducing it on my end.

Commit:
Cleans old code related to the placeholder Store.

- Removes unnecessary caching we did on this Store object before we had
Realm.io in place.

Commit:
Fixes a potential error with a Realm schema

Bug might have caused some issues on new installs.
